python/ray/tune/trial_executor.py

```diff
@@ -237,17 +266,28 @@ def force_reconcilation_on_next_step_end(self) -> None:
         pass
 
     def _may_warn_insufficient_resources(self, all_trials):
-        if not any(trial.status == Trial.RUNNING for trial in all_trials):
+        # This is approximately saying we are not making progress.
+        if len(all_trials) == self._all_trials_size:
             if self._no_running_trials_since == -1:
                 self._no_running_trials_since = time.monotonic()
             elif time.monotonic(
-            ) - self._no_running_trials_since > _get_warning_threshold():
-                # TODO(xwjiang): We should ideally output a more helpful msg.
+            ) - self._no_running_trials_since > _get_insufficient_resources_warning_threshold(  # noqa
+            ):
+                if not is_ray_cluster():  # autoscaler not enabled
+                    # If any of the pending trial cannot be fulfilled,
+                    # that's a good enough hint of trial resources not enough.
+                    if any(trial.status == Trial.PENDING
+                           and not _can_fulfill_no_autoscaler(trial)
+                           for trial in all_trials):
+                        raise TuneError(
+                            _get_insufficient_resources_warning_msg())
+                # TODO(xwjiang): Output a more helpful msg for autoscaler case.
                 # https://github.com/ray-project/ray/issues/17799
```

Comment:
Add a test for both these cases?

Reply:
Updated the tests. One specifically for verifying the TuneError raised.
Not sure how to cover autoscaler case tho. There doesn't seem to be a good way of mocking it.